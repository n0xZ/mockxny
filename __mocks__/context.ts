import { CartContextI } from 'types';
import { mockList } from './mocks';
export const mockDefaultValues: CartContextI = {
	clientCredit: 0,
	selectedProducts: [],
	addSelectedProduct: () => {},
	getTotalPrice: () => 0,
	updatePaymentStates: () => {},
	performPayment: () => {},
	removeSelectedProduct: () => {},
};

export const mockedSelectedProductValues: CartContextI = {
	clientCredit: 0,
	selectedProducts: mockList,
	addSelectedProduct: () => {},
	getTotalPrice: () => 0,
	updatePaymentStates: () => {},
	performPayment: () => {},
	removeSelectedProduct: () => {},
};
