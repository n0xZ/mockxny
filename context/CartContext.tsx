import {
	createContext,
	FC,
	useContext,
	useState,
} from 'react';
import { useRouter } from 'next/router';
import { CartContextI, Product } from '../types';
import { toast } from 'react-hot-toast';
// Cart provider initialized, with its values based on the CartContextI interface
export const CartProvider = createContext<CartContextI>({
	clientCredit: 0,
	selectedProducts: [],
	addSelectedProduct: () => {},
	removeSelectedProduct: () => {},
	updatePaymentStates: () => {},
	getTotalPrice: () => 0,
	performPayment: () => {},
});

export const CartContext: FC = ({ children }) => {
	const router = useRouter();

	const [clientCredit, setClientCredit] = useState(1000);
	const [selectedProducts, setSelectedProducts] = useState<
		Product[]
	>([]);
	const addSelectedProduct = (product: Product) => {
		toast.success('Producto agregado al carrito');
		setSelectedProducts([...selectedProducts, product]);
	};
	const removeSelectedProduct = (index: number) => {
		setSelectedProducts(
			selectedProducts.filter((p, ind) => ind !== index)
		);
	};
	const getTotalPrice = () => {
		return Math.round(
			selectedProducts.reduce(
				(prevValue, currentValue) =>
					prevValue + currentValue.price,
				0
			)
		);
	};
	const updatePaymentStates = () => {
		setClientCredit(clientCredit - getTotalPrice());
		setSelectedProducts([]);
	};
	const performPayment = () => {
		if (getTotalPrice() < clientCredit) {
			router.push('/cart/payment/successful');
		} else {
			router.push('/cart/payment/error');
		}
	};
	return (
		<CartProvider.Provider
			value={{
				clientCredit,
				selectedProducts,
				addSelectedProduct,
				updatePaymentStates,
				removeSelectedProduct,
				getTotalPrice,
				performPayment,
			}}
		>
			{children}
		</CartProvider.Provider>
	);
};

export const useCard = () => useContext(CartProvider);
