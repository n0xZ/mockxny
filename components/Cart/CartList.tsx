import React from 'react';
import { useCard } from '../../context/CartContext';
import Button from '../Button';
import CardItem from './CartItem';
import EmptyCart from './EmptyCart';
import CartClientCredit from './CartClientCredit';
import CartFinalPrice from './CartFinalPrice';
import Link from 'next/link';
const CartList = () => {
	const {
		selectedProducts,
		clientCredit,
		getTotalPrice,
		performPayment,
	} = useCard();
	return (
		<>
			{selectedProducts.length === 0 ? (
				<EmptyCart />
			) : (
				<article className="flex flex-col justify-between p-2">
					<h2 className="text-center text-3xl  mt-2 font-bold mb-2">
						Mis productos
					</h2>
					<div className='flex flex-col container mx-auto mt-6"'>
						{selectedProducts.map((product, index) => (
							<CardItem
								product={product}
								productIndex={index}
								key={index}
							/>
						))}
					</div>

					<div className="mt-4 flex flex-col">
						<CartFinalPrice finalPrice={getTotalPrice()} />
						<CartClientCredit clientAmount={clientCredit} />
						<div className=" mt-3 container mx-auto flex flex-row justify-between w-full  items-end">
							<Link href="/products">
								<a>Volver atr√°s</a>
							</Link>
							<Button onClick={() => performPayment()}>
								Realizar compra
							</Button>
						</div>
					</div>
				</article>
			)}
		</>
	);
};

export default CartList;
