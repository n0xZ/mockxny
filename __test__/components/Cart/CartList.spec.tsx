import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import CartList from '@/components/Cart/CartList';
import {
	mockDefaultValues,
	mockedSelectedProductValues,
} from '@/__mocks__/context';
import { CartProvider } from '@/context/CartContext';
describe('Cart list component test case', () => {
	it('Renders the component without any trouble', () => {
		const { container } = render(<CartList />);

		expect(container).toBeInTheDocument();
	});
	it(`Displays the empty cart component if there aren't any selected products`, () => {
		render(
			<CartProvider.Provider value={mockDefaultValues}>
				<CartList />
			</CartProvider.Provider>
		);
		const text = screen.getByText('Su carrito está vacío');
		expect(text).toBeInTheDocument();
	});
	it(`Displays the cart list component with its selected products`, () => {
		const { container } = render(
			<CartProvider.Provider
				value={mockedSelectedProductValues}
			>
				<CartList />
			</CartProvider.Provider>
		);
		const text = screen.getByText(`John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet`);
		expect(text).toBeInTheDocument()
		expect(container).toMatchSnapshot();
	});
});
