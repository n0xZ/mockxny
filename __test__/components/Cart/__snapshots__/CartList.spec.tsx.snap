// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart list component test case Displays the cart list component with its selected products 1`] = `
<div>
  <article
    class="flex flex-col justify-between p-2"
  >
    <h2
      class="text-center text-3xl mt-2 font-bold mb-2"
    >
      Mis productos
    </h2>
    <div
      class="flex flex-col container mx-auto mt-6\\""
    >
      <aside
        class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
      >
        <div
          class="flex flex-row items-center"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="Mens Casual Slim Fit"
              class="object-contain "
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
          <p
            class="font-bold mx-2 text-sm"
          >
            Mens Casual Slim Fit
          </p>
        </div>
        <div
          class="flex flex-row items-center"
        >
          <p
            class="mx-2 text-sm"
          >
            $ 15.99
          </p>
          <button
            class="border-2 font-lato font-bold border-black rounded-md px-2 hover:border-2 hover:bg-gray-200 duration-100 ease-in"
          >
            X
          </button>
        </div>
      </aside>
      <aside
        class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
      >
        <div
          class="flex flex-row items-center"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet"
              class="object-contain "
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
          <p
            class="font-bold mx-2 text-sm"
          >
            John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet
          </p>
        </div>
        <div
          class="flex flex-row items-center"
        >
          <p
            class="mx-2 text-sm"
          >
            $ 695
          </p>
          <button
            class="border-2 font-lato font-bold border-black rounded-md px-2 hover:border-2 hover:bg-gray-200 duration-100 ease-in"
          >
            X
          </button>
        </div>
      </aside>
      <aside
        class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
      >
        <div
          class="flex flex-row items-center"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="Solid Gold Petite Micropave "
              class="object-contain "
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
          <p
            class="font-bold mx-2 text-sm"
          >
            Solid Gold Petite Micropave 
          </p>
        </div>
        <div
          class="flex flex-row items-center"
        >
          <p
            class="mx-2 text-sm"
          >
            $ 168
          </p>
          <button
            class="border-2 font-lato font-bold border-black rounded-md px-2 hover:border-2 hover:bg-gray-200 duration-100 ease-in"
          >
            X
          </button>
        </div>
      </aside>
      <aside
        class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
      >
        <div
          class="flex flex-row items-center"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="White Gold Plated Princess"
              class="object-contain "
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
          <p
            class="font-bold mx-2 text-sm"
          >
            White Gold Plated Princess
          </p>
        </div>
        <div
          class="flex flex-row items-center"
        >
          <p
            class="mx-2 text-sm"
          >
            $ 9.99
          </p>
          <button
            class="border-2 font-lato font-bold border-black rounded-md px-2 hover:border-2 hover:bg-gray-200 duration-100 ease-in"
          >
            X
          </button>
        </div>
      </aside>
      <aside
        class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
      >
        <div
          class="flex flex-row items-center"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="Pierced Owl Rose Gold Plated Stainless Steel Double"
              class="object-contain "
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
          <p
            class="font-bold mx-2 text-sm"
          >
            Pierced Owl Rose Gold Plated Stainless Steel Double
          </p>
        </div>
        <div
          class="flex flex-row items-center"
        >
          <p
            class="mx-2 text-sm"
          >
            $ 10.99
          </p>
          <button
            class="border-2 font-lato font-bold border-black rounded-md px-2 hover:border-2 hover:bg-gray-200 duration-100 ease-in"
          >
            X
          </button>
        </div>
      </aside>
    </div>
    <div
      class="mt-4 flex flex-col"
    >
      <div
        class="h-20 flex flex-row container mx-auto mt-6\\""
      >
        <aside
          class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
        >
          <h2>
            Precio Final
          </h2>
          <span
            class="flex flex-row"
          >
            <p
              class="font-bold mx-2"
            >
              $ 
              0
            </p>
          </span>
        </aside>
      </div>
      <div
        class="h-20 flex flex-row container mx-auto mt-6\\""
      >
        <aside
          class="flex flex-row justify-between items-center rounded-md w-full p-5 border-gray-50 border-2 mt-2 hover:opacity-90 duration-100 ease-in-out"
        >
          <h2>
            Tu saldo
          </h2>
          <span
            class="flex flex-row"
          >
            <p
              class="font-bold mx-2"
            >
              $ 
              0
            </p>
          </span>
        </aside>
      </div>
      <div
        class=" mt-3 container mx-auto flex flex-row justify-between w-full items-center"
      >
        <a
          href="/products"
        >
          Volver atrás
        </a>
        <button
          class="undefined border-2 font-lato text-base border-black rounded-md px-6 py-4 hover:border-gray-900 hover:bg-gray-200  duration-100 ease-in"
        >
          Realizar compra
        </button>
      </div>
    </div>
  </article>
</div>
`;
